compiler/rustc_trait_selection/src/solve/search_graph/mod.rs

```diff
@@ -204,13 +205,18 @@ impl<'tcx> SearchGraph<'tcx> {
                 let actual_index = cache.lookup_table.remove(&entry.goal);
                 debug_assert_eq!(Some(i), actual_index);
                 debug_assert!(entry.depth == depth);
-                cache::try_move_finished_goal_to_global_cache(
-                    tcx,
-                    &mut self.overflow_data,
-                    &self.stack,
-                    entry.goal,
-                    entry.response,
-                );
+
+                // FIXME: ???
+                if i == provisional_entry_index.index() {
```

Comment:
Is this the proper way of checking the head of the cycle? Just the first entry in this drain iterator?

(modulo the type error that currently exists)

Reply:
jup :+1: though it's probably nicer to first drain all other cycle participants. SO something like
```rust
            // If the current goal is the head of a cycle, we drop all other
            // cycle participants without moving them to the global cache.
            let other_cycle_participants = provisional_entry_index.index() + 1;
            for (i, entry) in cache.entries.drain_enumerated(other_cycle_participants..) {
                let actual_index = cache.lookup_table.remove(&entry.goal);
                debug_assert_eq!(Some(i), actual_index);
                debug_assert!(entry.depth == depth);
            }
            
            let current_goal = cache.entries.pop().unwrap();
            let actual_index = cache.lookup_table.remove(&entry.goal);
            debug_assert_eq!(Some(i), provisional_entry_index.index());
            debug_assert!(entry.depth == depth);
            try_to_global_cache
```
can we somehow move the pop `entries` and fix `lookup_table` into a function :thinking: it's a bit scary that we may forget to do one of them somewhere

Reply:
> can we somehow move the pop entries and fix lookup_table into a function

Hm... I can't see how that works with the `drain_enumerated`, unless we'd write a wrapper around that or something?